crates/red_knot/src/symbols.rs

```diff
@@ -860,13 +880,83 @@ impl PreorderVisitor<'_> for SymbolTableBuilder {
                 ast::visitor::preorder::walk_stmt(self, stmt);
                 self.current_definition = None;
             }
+            ast::Stmt::If(node) => {
+                self.visit_expr(&node.test);
+                let pre_body = self.current_flow_node();
+                let if_branch = self.new_flow_node(FlowNode::Branch(BranchFlowNode {
+                    predecessor: pre_body,
+                }));
+                self.set_current_flow_node(if_branch);
+                self.visit_body(&node.body);
+                let mut phi_node = PhiFlowNode {
+                    predecessors: vec![self.current_flow_node()],
+                };
+                let mut last_branch = if_branch;
+                let mut last_branch_is_else = false;
+                for clause in &node.elif_else_clauses {
+                    if clause.test.is_some() {
+                        let clause_branch = self.new_flow_node(FlowNode::Branch(BranchFlowNode {
+                            predecessor: last_branch,
+                        }));
+                        last_branch = clause_branch;
+                        self.set_current_flow_node(clause_branch);
+                    } else {
+                        self.set_current_flow_node(last_branch);
+                        last_branch_is_else = true;
+                    }
+                    self.visit_elif_else_clause(clause);
+                    phi_node.predecessors.push(self.current_flow_node());
+                }
+                if !last_branch_is_else {
+                    phi_node.predecessors.push(last_branch);
+                }
+                let phi_node_id = self.new_flow_node(FlowNode::Phi(phi_node));
+                self.set_current_flow_node(phi_node_id);
```

Comment:
This is a nit, but I think adding some empty lines to group the logic would help readers to better understand what's happening. It now reads like a text without any paragraphs, which makes it hard to scan.
```suggestion
                self.visit_expr(&node.test);

                let pre_body = self.current_flow_node();
                let if_branch = self.new_flow_node(FlowNode::Branch(BranchFlowNode {
                    predecessor: pre_body,
                }));
                
                self.set_current_flow_node(if_branch);
                self.visit_body(&node.body);
                
                let mut phi_node = PhiFlowNode {
                    predecessors: vec![self.current_flow_node()],
                };
                let mut last_branch = if_branch;
                let mut last_branch_is_else = false;
                
                for clause in &node.elif_else_clauses {
                    if clause.test.is_some() {
                        let clause_branch = self.new_flow_node(FlowNode::Branch(BranchFlowNode {
                            predecessor: last_branch,
                        }));
                        last_branch = clause_branch;
                        self.set_current_flow_node(clause_branch);
                    } else {
                        self.set_current_flow_node(last_branch);
                        last_branch_is_else = true;
                    }
                    self.visit_elif_else_clause(clause);
                    phi_node.predecessors.push(self.current_flow_node());
                }
                
                if !last_branch_is_else {
                    phi_node.predecessors.push(last_branch);
                }
                
                let phi_node_id = self.new_flow_node(FlowNode::Phi(phi_node));
                self.set_current_flow_node(phi_node_id);
```

Reply:
Good call; I did this, and also added comments explaining the algorithm.